engine/lib/phx/src/input/devices/mouse.rs

```diff
@@ -0,0 +1,161 @@
+use std::collections::HashMap;
+
+use glam::Vec2;
+
+use crate::{input::*, system::TimeStamp};
+
+#[luajit_ffi_gen::luajit_ffi]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MouseControl {
+    Left,
+    Middle,
+    Right,
+    X1,
+    X2,
+    DeltaX,
+    DeltaY,
+    ScrollX,
+    ScrollY,
+    ScrollPixelX,
+    ScrollPixelY,
+}
+
+impl MouseControl {
+    pub fn is_button(&self) -> bool {
+        match self {
+            Self::Left | Self::Middle | Self::Right | Self::X1 | Self::X2 => true,
+            _ => false,
+        }
+    }
+
+    pub fn is_axis(&self) -> bool {
+        match self {
+            Self::DeltaX
+            | Self::DeltaY
+            | Self::ScrollX
+            | Self::ScrollY
+            | Self::ScrollPixelX
+            | Self::ScrollPixelY => true,
+            _ => false,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct MouseState {
+    control_state: ControlState,
+    button_state: ButtonState<{ MouseControl::SIZE }>,
```

Comment:
I've never seen this syntax before. Why can't we just write the below?
```suggestion
    button_state: ButtonState<MouseControl::SIZE>,
```

Reply:
Already tried your suggestion before, it doesn't work. Looks like it's a standard way to specify non-literal generic consts.

Reply:
interesting, fair enough, must be an oddity with the parser.